util/types/convert.go

```diff
@@ -138,78 +147,151 @@ func isCastType(tp byte) bool {
 }
 
 // StrToInt converts a string to an integer in best effort.
-// TODO: handle overflow and add unittest.
 func StrToInt(sc *variable.StatementContext, str string) (int64, error) {
 	str = strings.TrimSpace(str)
-	if len(str) == 0 {
-		return 0, nil
-	}
-	negative := false
-	i := 0
-	if str[i] == '-' {
-		negative = true
-		i++
-	} else if str[i] == '+' {
-		i++
-	}
-	r := int64(0)
-	for ; i < len(str); i++ {
-		if !unicode.IsDigit(rune(str[i])) {
-			break
+	validPrefix, err := getValidIntPrefix(sc, str)
+	iVal, err1 := strconv.ParseInt(validPrefix, 10, 64)
+	if err1 != nil {
+		return iVal, errors.Trace(ErrOverflow)
+	}
+	return iVal, errors.Trace(err)
+}
+
+// StrToUint converts a string to an unsigned interger in best effort.
+func StrToUint(sc *variable.StatementContext, str string) (uint64, error) {
+	str = strings.TrimSpace(str)
+	validPrefix, err := getValidIntPrefix(sc, str)
+	if validPrefix[0] == '+' {
+		validPrefix = validPrefix[1:]
+	}
+	uVal, err1 := strconv.ParseUint(validPrefix, 10, 64)
+	if err1 != nil {
+		return uVal, errors.Trace(ErrOverflow)
+	}
+	return uVal, err
+}
+
+// getValidIntPrefix gets prefix of the string which can be successfully parsed as int.
+func getValidIntPrefix(sc *variable.StatementContext, str string) (string, error) {
+	floatPrefix, err := getValidFloatPrefix(sc, str)
+	if err != nil {
+		return floatPrefix, errors.Trace(err)
+	}
+	return floatStrToIntStr(floatPrefix)
+}
+
+// floatStrToIntStr converts a valid float string into valid integer string which can be parsed by
+// strconv.ParseInt, we can't parse float first then convert it to string because precision will
+// be lost.
+func floatStrToIntStr(validFloat string) (string, error) {
+	var dotIdx = -1
+	var eIdx = -1
+	for i := 0; i < len(validFloat); i++ {
+		switch validFloat[i] {
+		case '.':
+			dotIdx = i
+		case 'e', 'E':
+			eIdx = i
 		}
-		r = r*10 + int64(str[i]-'0')
 	}
-	if negative {
-		r = -r
+	if dotIdx == -1 && eIdx == -1 {
```

Comment:
Then it is not valid float string.
This function only handle valid float string.